iox_catalog/src/lib.rs

```diff
@@ -149,47 +150,41 @@ where
     // If the table itself needs to be updated during column validation it
     // becomes a Cow::owned() copy and the modified copy should be inserted into
     // the schema before returning.
-    let column_batch: Vec<_> = mb
-        .columns()
-        .filter_map(|(name, col)| {
-            // Check if the column exists in the cached schema.
-            //
-            // If it does, validate it. If it does not exist, create it and insert
-            // it into the cached schema.
-            match table.columns.get(name.as_str()) {
-                Some(existing) if existing.matches_type(col.influx_type()) => {
-                    // No action is needed as the column matches the existing column
-                    // schema.
-                    None
-                }
-                Some(existing) => {
-                    // The column schema, and the column in the mutable batch are of
-                    // different types.
-                    Some(
-                        ColumnTypeMismatchSnafu {
-                            name,
-                            existing: existing.column_type,
-                            new: col.influx_type(),
-                        }
-                        .fail(),
-                    )
-                }
-                None => {
-                    // The column does not exist in the cache, add it to the column
-                    // batch to be bulk inserted later.
-                    Some(Ok(ColumnUpsertRequest {
-                        name: name.as_str(),
-                        column_type: ColumnType::from(col.influx_type()),
-                    }))
+    let mut column_batch: HashMap<&str, ColumnType> = HashMap::new();
+
+    for (name, col) in mb.columns() {
+        // Check if the column exists in the cached schema.
+        //
+        // If it does, validate it. If it does not exist, create it and insert
+        // it into the cached schema.
+
+        match table.columns.get(name.as_str()) {
+            Some(existing) if existing.matches_type(col.influx_type()) => {
+                // No action is needed as the column matches the existing column
+                // schema.
+            }
+            Some(existing) => {
+                // The column schema, and the column in the mutable batch are of
+                // different types.
+                return ColumnTypeMismatchSnafu {
+                    name,
+                    existing: existing.column_type,
+                    new: col.influx_type(),
                 }
+                .fail();
             }
-        })
-        .collect::<Result<Vec<_>>>()?;
+            None => {
+                // The column does not exist in the cache, add it to the column
+                // batch to be bulk inserted later.
+                column_batch.insert(name.as_str(), ColumnType::from(col.influx_type()));
```

Comment:
Perhaps this is a bit excessive, but I usually assert this wasn't an _update_ despite the method name (it always gets me!):

```suggestion
                let old = column_batch.insert(name.as_str(), ColumnType::from(col.influx_type()));
                assert!(old.is_none(), "duplicate column name in new column batch");
```

Replacing an existing entry **should** never happen with the current code 🤞, and the old code perhaps should have checked there were no duplicates too. With the new hashmap approach this check is _cheap_ though 🎉 

(this could also be applied to the TSM merging logic if you fancy)